# Installing pandas

In [ ]:
!pip install pandas

#Now let's import it



In [46]:
import pandas as pd

#Series and Dataframes
#  What is a Series?


A **Series** is a one-dimensional labeled array capable of holding any data type (integers, strings, floats, Python objects, etc).

Think of it as:
- A **single column** from a table
- A **list** with labels (called **index**)

In [47]:
data = [100, 200, 300]
s = pd.Series(data,name = 'series_1')

print("Series:\n", s)


Series:
 0    100
1    200
2    300
Name: series_1, dtype: int64


In [48]:
index = ["a", "b", "c"]
data = [10,20,30]
s = pd.Series(data, index=index) #index => identifier for rows

print("Series:\n", s)

Series:
 a    10
b    20
c    30
dtype: int64


## What is a DataFrame?

A **DataFrame** is a 2-dimensional table (like an Excel sheet or SQL table) with rows and columns.

- Think of it as a **collection of Series**
- Each column is a Series
- Rows and columns can have labels

In [49]:
data = {
    "Name": ["Alice", "Bob", "Charlie"],
    "Age": [25, 30, 35],
    "City": ["NYC", "LA", "Chicago"]  #each key of a dict represent a column name and each element of a list represent a row
}

index = ["क", "ख", "ग"] #doesnt need to be unique but generally is made unique

df = pd.DataFrame(data,index)
print("DataFrame:\n", df)

DataFrame:
       Name  Age     City
क    Alice   25      NYC
ख      Bob   30       LA
ग  Charlie   35  Chicago


In [50]:
data = [
    ["Alice", 25],
    ["Bob", 30],
    ["Charlie", 35]
]   #each element of the outer list represent a row and inner list represent column

df = pd.DataFrame(data, columns=["Name", "Age"])
print(df)

      Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35


In [51]:
data = [
    {"Name": "Alice", "Age": 25},
    {"Name": "Bob", "Age": 30},
    {"Name": "Charlie", "Age": 35}
]   #each element of the list represent a row and dict key:name of the column value:value of that column

df = pd.DataFrame(data)
print(df)

      Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35


In [52]:
s1 = pd.Series([25, 30, 35])
s2 = pd.Series(["Alice", "Bob", "Charlie"])


df = pd.DataFrame({
    "Name": s2,
    "Age": s1   #each series represent the data for a column
})

print(df)


      Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35


In [53]:
person1 = pd.Series(["Alice", 25, "F"], index=["name", "age", "sex"])
person2 = pd.Series(["Bob", 30, "M"], index=["name", "age", "sex"])
person3 = pd.Series(["Charlie", 22, "M"], index=["name", "age", "sex"])

df = pd.DataFrame([person1, person2, person3])   #each series represent a row

print(df)



      name  age sex
0    Alice   25   F
1      Bob   30   M
2  Charlie   22   M


# Accessing columns
We can use df['column name'] to access columns of a dataframe

In [54]:
df['name']

0      Alice
1        Bob
2    Charlie
Name: name, dtype: object

In [55]:
type(df['name'])

pandas.core.series.Series

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using.

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable.

The dataset contains the following columns:

Daynum: The day of the season the game was played.

Wteam: The team ID of the winning team.

Wscore: The score of the winning team.

Lteam: The team ID of the losing team.

Lscore: The score of the losing team.

Wloc: The location where the game was played (H for Home, A for Away, N for Neutral)

Numot: The number of overtime periods played.

In [128]:
df = pd.read_csv('RegularSeasonCompactResults.csv')

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [129]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228    81.0   1328    64.0    N    0.0
1    1985      25   1106    77.0   1354     NaN    H    NaN
2    1985      25   1112     NaN   1223     NaN    H    0.0
3    1985      25   1165    70.0   1432    54.0    H    0.0
4    1985      25   1192    86.0   1447    74.0    H    0.0

In [130]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114    70.0   1419    50.0    N    0.0
145285    2016     132   1163    72.0   1272    58.0    N    0.0
145286    2016     132   1246    82.0   1401    77.0    N    1.0
145287    2016     132   1277    66.0   1345    62.0    N    0.0
145288    2016     132   1386    87.0   1433    74.0    N    0.0

We can see the dimensions of the dataframe using the the **shape** attribute

In [131]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [132]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset.

In [133]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  130716.000000   
mean     2001.574834      75.223816    1286.720646      76.599858   
std         9.233342      33.287418     104.570275      12.180251   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  130715.000000  130761.000000  
mean     1282.864064      64.486463       0.044486  
std       104.829234      11.389853       0.248061  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Let's see what info() does

In [134]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145289 entries, 0 to 145288
Data columns (total 8 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Season  145289 non-null  int64  
 1   Daynum  145289 non-null  int64  
 2   Wteam   145289 non-null  int64  
 3   Wscore  130716 non-null  float64
 4   Lteam   145289 non-null  int64  
 5   Lscore  130715 non-null  float64
 6   Wloc    145289 non-null  object 
 7   Numot   130761 non-null  float64
dtypes: float64(3), int64(4), object(1)
memory usage: 8.9+ MB


Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [135]:
df.max()

Season     2016
Daynum      132
Wteam      1464
Wscore    186.0
Lteam      1464
Lscore    150.0
Wloc          N
Numot       6.0
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [136]:
df['Wscore'].max()

186.0

If you'd like to find the mean of the Losing teams' score.

In [137]:
df['Lscore'].mean()

np.float64(64.4864629155032)

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [138]:
df['Wscore'].argmax()


np.int64(24970)

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [139]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** indexer. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [140]:
print(df.iloc[24970])

Season     1991
Daynum       68
Wteam      1258
Wscore    186.0
Lteam      1109
Lscore    140.0
Wloc          H
Numot       0.0
Name: 24970, dtype: object


Note: Iloc is not a function!!! Its an object which implements an indexer class (having a \_\_getitem__ function). Let's look at a simple class having \_\_getitem__.





In [141]:
class dummyindexer:
    def __init__(self, data):
        self.data = data

    def __getitem__(self, key):
        print(f"Getting item at: {key}")
        if isinstance(key, tuple):
            row, col = key
            return self.data[row][col]
        else:
            return self.data[key]

data = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

obj = dummyindexer(data)

print(obj[1])  # This calls obj.__getitem__(1) behind the scenes
print(obj[1, 2])    #This calls obj.__getitem((1,2))
print(obj.__getitem__((1,2)))


Getting item at: 1
[4, 5, 6]
Getting item at: (1, 2)
6
Getting item at: (1, 2)
6


In [142]:
print(df.iloc.__getitem__(24970))  #this gets called under the hood

Season     1991
Daynum       68
Wteam      1258
Wscore    186.0
Lteam      1109
Lscore    140.0
Wloc          H
Numot       0.0
Name: 24970, dtype: object


Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored along with the daynum.

In [143]:
print(df.iloc[24970][['Lscore','Daynum']])
#or
print('\n\n')

print(df.iloc[24970, [5,1]])   #this only accepts integers

Lscore    140.0
Daynum       68
Name: 24970, dtype: object



Lscore    140.0
Daynum       68
Name: 24970, dtype: object


When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [144]:
type(df.iloc[24970])


pandas.core.series.Series

In [145]:
print(df.iloc[24970].index)

Index(['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc',
       'Numot'],
      dtype='object')


The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [146]:
df.iloc[24970]['Lscore']

np.float64(140.0)

The other really important object in Pandas is **loc** . Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [147]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228    81.0   1328    64.0    N    0.0
1    1985      25   1106    77.0   1354     NaN    H    NaN
2    1985      25   1112     NaN   1223     NaN    H    0.0

In [148]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228    81.0   1328    64.0    N    0.0
1    1985      25   1106    77.0   1354     NaN    H    NaN
2    1985      25   1112     NaN   1223     NaN    H    0.0
3    1985      25   1165    70.0   1432    54.0    H    0.0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive.

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [149]:
df.loc[24970, 'Lscore']

np.float64(140.0)

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get.

In [150]:
df.at[24970, 'Lscore']

np.float64(140.0)

Let's look at a better example

In [151]:
data = {
    "Name": ["Alice", "Bob", "Charlie","Chris"],
    "Age": [25, 30, 35,32],
    "City": ["NYC", "LA", "Chicago","NYC"]
}

index = ["क", "ख","⭐" ,"ख"]     #making the indexes non unique for demonstration purposes

example_dataframe = pd.DataFrame(data,index)
print("DataFrame:\n", example_dataframe)

DataFrame:
       Name  Age     City
क    Alice   25      NYC
ख      Bob   30       LA
⭐  Charlie   35  Chicago
ख    Chris   32      NYC


In [152]:
example_dataframe.iloc[0]

Name    Alice
Age        25
City      NYC
Name: क, dtype: object

In [153]:
example_dataframe.loc["क"]

Name    Alice
Age        25
City      NYC
Name: क, dtype: object

In [154]:
example_dataframe.loc['ख']

Name  Age City
ख    Bob   30   LA
ख  Chris   32  NYC

In [155]:
example_dataframe.loc["क":'⭐']

Name  Age     City
क    Alice   25      NYC
ख      Bob   30       LA
⭐  Charlie   35  Chicago

In [156]:
example_dataframe.loc["क":"ख"] #doesn't work because of non-unique index (but if the non unique indexes were together, it would work)

KeyError: "Cannot get right slice bound for non-unique label: 'ख'"

In [157]:
example_dataframe.iloc[0:3]

Name  Age     City
क    Alice   25      NYC
ख      Bob   30       LA
⭐  Charlie   35  Chicago

# Null values
We will encounter several missing values in our data. The .isnull() method in pandas checks each value in a DataFrame or Series and returns:


  True if the value is null (i.e., NaN, None, or missing)

  False if the value is not null

In [158]:
df.isnull()   #returns a dataframe containing true/false

Season  Daynum  Wteam  Wscore  Lteam  Lscore   Wloc  Numot
0        False   False  False   False  False   False  False  False
1        False   False  False   False  False    True  False   True
2        False   False  False    True  False    True  False  False
3        False   False  False   False  False   False  False  False
4        False   False  False   False  False   False  False  False
...        ...     ...    ...     ...    ...     ...    ...    ...
145284   False   False  False   False  False   False  False  False
145285   False   False  False   False  False   False  False  False
145286   False   False  False   False  False   False  False  False
145287   False   False  False   False  False   False  False  False
145288   False   False  False   False  False   False  False  False

[145289 rows x 8 columns]

In [159]:
df.isnull().sum(axis=0) #sum the values of that dataframe (true=1 false =0)

Season        0
Daynum        0
Wteam         0
Wscore    14573
Lteam         0
Lscore    14574
Wloc          0
Numot     14528
dtype: int64

For learning purposes, let's see what this does

In [160]:
df.isnull().sum(axis=1)

0         0
1         2
2         2
3         0
4         0
         ..
145284    0
145285    0
145286    0
145287    0
145288    0
Length: 145289, dtype: int64

We can also fill values in null entries

In [161]:
df['Wscore']=df['Wscore'].fillna(df['Wscore'].mean())    #filling the null values of these columns with the mean of the column
df['Lscore'] = df['Lscore'].fillna(df['Lscore'].mean())

For now, lets just delete the rows having null value in any column

In [162]:
df.dropna(inplace=True)  #inplace = true replaces that dataframe with the new dataframe (equivalent to df=df.dropna())


# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [163]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
89021     2006      44   1284    41.0   1343    21.0    A    0.0
49310     1997      66   1157    61.0   1204    21.0    H    0.0
103660    2009      26   1326    59.0   1359    22.0    H    0.0
85042     2005      66   1131    73.0   1216    22.0    H    0.0
88130     2006      20   1332    83.0   1366    23.0    H    0.0

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, we want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [164]:
df['Wscore'] > 150   #returns a boolean series containing true/false

0         False
2         False
3         False
4         False
5         False
          ...  
145284    False
145285    False
145286    False
145287    False
145288    False
Name: Wscore, Length: 130761, dtype: bool

In [165]:
df[df['Wscore']>150]   #we can use the mask to filter the dataframe

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258   151.0   1109   107.0    H    0.0
12046    1988      40   1328   152.0   1147    84.0    H    0.0
12355    1988      52   1328   151.0   1173    99.0    N    0.0
16040    1989      40   1328   152.0   1331   122.0    H    0.0
16853    1989      68   1258   162.0   1109   144.0    A    0.0
19653    1990      30   1328   173.0   1109   101.0    H    0.0
19971    1990      38   1258   152.0   1109   137.0    A    0.0
20022    1990      40   1116   166.0   1109   101.0    H    0.0
23582    1991      26   1318   152.0   1258   123.0    N    0.0
24341    1991      47   1328   172.0   1258   112.0    H    0.0
24970    1991      68   1258   186.0   1109   140.0    H    0.0
28687    1992      54   1261   159.0   1319    86.0    H    0.0
35023    1993     112   1380   155.0   1341    91.0    A    0.0
40060    1995      32   1375   156.0   1341   114.0    H    0.0
52600    1998      33   1395   153.0   1410    87.0    H    0.0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100.

In [166]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328   152.0   1147    84.0    H    0.0
12355    1988      52   1328   151.0   1173    99.0    N    0.0
28687    1992      54   1261   159.0   1319    86.0    H    0.0
35023    1993     112   1380   155.0   1341    91.0    A    0.0
52600    1998      33   1395   153.0   1410    87.0    H    0.0

#Map and apply

In pandas, `map()` and `apply()` are used to transform data by applying custom functions but they work differently and serve different purposes.

  `map()`
- Used with **Series** only.
- Applies a function **element-wise** to each value.
- Great for simple transformations.

In [167]:
transformed_wscore = df['Wscore'].map(lambda x: x/255)
# lambda syntax: lambda input: expression
# Each element in the Series is passed to the lambda function, and the result is returned in a new Series.

transformed_wscore

0         0.317647
2         0.300392
3         0.274510
4         0.337255
5         0.309804
            ...   
145284    0.274510
145285    0.282353
145286    0.321569
145287    0.258824
145288    0.341176
Name: Wscore, Length: 130761, dtype: float64

 `apply()`

  - Can be used with Series or DataFrame.

  - For Series: behaves like map().

  - For DataFrames: can apply functions row-wise or column-wise using the axis argument.

In [168]:
score_diff = df.apply(lambda row: row['Wscore'] - row['Lscore'],axis=1)  #axis = 1 because axis=0 passes columns but we need to work with rows
score_diff

0         17.000000
2         12.113395
3         16.000000
4         12.000000
5          1.000000
            ...    
145284    20.000000
145285    14.000000
145286     5.000000
145287     4.000000
145288    13.000000
Length: 130761, dtype: float64

Adding the new series to our dataframe

In [169]:
df['Difference']= score_diff
df.head()


Season  Daynum  Wteam     Wscore  Lteam     Lscore Wloc  Numot  Difference
0    1985      20   1228  81.000000   1328  64.000000    N    0.0   17.000000
2    1985      25   1112  76.599858   1223  64.486463    H    0.0   12.113395
3    1985      25   1165  70.000000   1432  54.000000    H    0.0   16.000000
4    1985      25   1192  86.000000   1447  74.000000    H    0.0   12.000000
5    1985      25   1218  79.000000   1337  78.000000    H    0.0    1.000000

#Vector operations and broadcasting

Vector operations in pandas allow you to perform arithmetic or logical operations on entire Series or DataFrames at once, without writing explicit loops. These operations happen element-wise, meaning that the operation is applied independently to each matching element based on the index labels.

Broadcasting is a related concept where pandas automatically expands smaller objects to match the shape of larger ones during operations. For instance, when you add a scalar value to a DataFrame, pandas "broadcasts" the scalar across every element in the DataFrame, effectively performing the operation element-wise without you having to write explicit code for it.

In [170]:
#We can calculate the differnce of scores by using vector operations
difference_by_vector_operations = df['Wscore'] - df['Lscore']
difference_by_vector_operations

0         17.000000
2         12.113395
3         16.000000
4         12.000000
5          1.000000
            ...    
145284    20.000000
145285    14.000000
145286     5.000000
145287     4.000000
145288    13.000000
Length: 130761, dtype: float64

In [171]:
#Let's divide the difference by 10
difference_by_vector_operations = difference_by_vector_operations/10 #this is also a vector operation. it broadcasts the scalar value so that all the elements are divided by 10
difference_by_vector_operations

0         1.700000
2         1.211339
3         1.600000
4         1.200000
5         0.100000
            ...   
145284    2.000000
145285    1.400000
145286    0.500000
145287    0.400000
145288    1.300000
Length: 130761, dtype: float64

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by season) and then perform operations on them. The following function groups all the entries (games) on the same season  and finds the mean for each group.

In [172]:
df.groupby('Season')['Difference'].mean()

Season
1985    11.065722
1986    11.332236
1987    11.376611
1988    12.134627
1989    12.303839
1990    11.900177
1991    12.481568
1992    12.823944
1993    12.187756
1994    12.604009
1995    12.649714
1996    12.327557
1997    12.576865
1998    12.724297
1999    12.497287
2000    12.373419
2001    12.352616
2002    12.273679
2003    12.126615
2004    11.960907
2005    12.112001
2006    11.969177
2007    12.028799
2008    12.194595
2009    12.082717
2010    12.038260
2011    11.861502
2012    12.027815
2013    11.917600
2014    11.692949
2015    11.720879
2016    12.042924
Name: Difference, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [173]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        11
       A         3
       N         2
1102   H       189
       A        67
       N        28
1103   H       295
       A       140
       N        36
Name: count, dtype: int64

Now, let's see the number of gaves won at home, away

In [174]:
df.groupby('Wloc').size()

Wloc
A    39805
H    77831
N    13125
dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [175]:
df.values

array([[1985, 20, 1228, ..., 'N', 0.0, 17.0],
       [1985, 25, 1112, ..., 'H', 0.0, 12.113394791280982],
       [1985, 25, 1165, ..., 'H', 0.0, 16.0],
       ...,
       [2016, 132, 1246, ..., 'N', 1.0, 5.0],
       [2016, 132, 1277, ..., 'N', 0.0, 4.0],
       [2016, 132, 1386, ..., 'N', 0.0, 13.0]], dtype=object)

Now, you can simply just access elements like you would in an array.

In [176]:
df.values[0][0]

1985

# Finding outliers
In this demonstration, we will use Interquartile range to find out the outliers.

- **Q1** = 25th percentile  
- **Q3** = 75th percentile  
- **IQR** = Q3 - Q1

Outliers are defined as data points that lie:

- **Below:** Q1 - 1.5 × IQR  
- **Above:** Q3 + 1.5 × IQR

Let's find the outliers in the 'Difference' column

In [177]:
Q1 = df['Difference'].quantile(0.25)
Q3 = df['Difference'].quantile(0.75)
IQR = Q3 - Q1
print('Q1: ',Q1)
print('Q3: ',Q3)

Q1:  5.0
Q3:  17.0


In [178]:
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

In [179]:
outliers = df[(df['Difference'] < lower_bound) | (df['Difference'] > upper_bound)]
print(outliers)

        Season  Daynum  Wteam  Wscore  Lteam     Lscore Wloc  Numot  \
16        1985      26   1120    92.0   1391  50.000000    H    0.0   
25        1985      26   1196   106.0   1416  55.000000    H    0.0   
35        1985      26   1286   109.0   1186  64.000000    H    0.0   
42        1985      26   1375   104.0   1126  68.000000    H    0.0   
51        1985      26   1443    93.0   1121  57.000000    H    0.0   
...        ...     ...    ...     ...    ...        ...  ...    ...   
144916    2016     122   1372    95.0   1322  55.000000    A    0.0   
145010    2016     124   1308    83.0   1410  44.000000    A    0.0   
145146    2016     128   1397    97.0   1120  59.000000    N    0.0   
145148    2016     128   1424   108.0   1102  64.486463    N    3.0   
145245    2016     130   1372   104.0   1223  68.000000    N    0.0   

        Difference  
16       42.000000  
25       51.000000  
35       45.000000  
42       36.000000  
51       36.000000  
...            ...  


There are many ways to handle outliers. For this demo, we will just cap the outliers to the threshold

In [180]:
df['Difference'] = df['Difference'].clip(lower=lower_bound, upper=upper_bound)


In [181]:
outliers = df[(df['Difference'] < lower_bound) | (df['Difference'] > upper_bound)]
print(outliers)

Empty DataFrame
Columns: [Season, Daynum, Wteam, Wscore, Lteam, Lscore, Wloc, Numot, Difference]
Index: []


#One hot encoding
One hot encoding is a technique used to convert **categorical variables** into a format that can be provided to machine learning algorithms.

It creates a **new binary column for each category**, where:

- 1 indicates the presence of that category
- 0 indicates its absence



## Why is it needed?

Machine learning models can only work with **numeric** data.

If you have categories like `"Red"`, `"Green"`, `"Blue"`:
- You can't feed them directly into models
- Assigning numbers like Red=0, Green=1, Blue=2 can mislead the model (it assumes **order** and **distance** that don’t exist)

One-hot encoding avoids this by representing categories as **independent features**.



## Example

Suppose you have a column `"Color"`:

| Color |
|-------|
| Red   |
| Blue  |
| Green |

One-hot encoding transforms it into:

| Color_Red | Color_Blue | Color_Green |
|-----------|------------|-------------|
|     1     |     0      |      0      |
|     0     |     1      |      0      |
|     0     |     0      |      1      |

---

In [182]:
one_hot_encodings = pd.get_dummies(df['Wloc'])
one_hot_encodings

A      H      N
0       False  False   True
2       False   True  False
3       False   True  False
4       False   True  False
5       False   True  False
...       ...    ...    ...
145284  False  False   True
145285  False  False   True
145286  False  False   True
145287  False  False   True
145288  False  False   True

[130761 rows x 3 columns]

In [183]:
df = pd.concat([df, one_hot_encodings], axis=1)
df

Season  Daynum  Wteam     Wscore  Lteam     Lscore Wloc  Numot  \
0         1985      20   1228  81.000000   1328  64.000000    N    0.0   
2         1985      25   1112  76.599858   1223  64.486463    H    0.0   
3         1985      25   1165  70.000000   1432  54.000000    H    0.0   
4         1985      25   1192  86.000000   1447  74.000000    H    0.0   
5         1985      25   1218  79.000000   1337  78.000000    H    0.0   
...        ...     ...    ...        ...    ...        ...  ...    ...   
145284    2016     132   1114  70.000000   1419  50.000000    N    0.0   
145285    2016     132   1163  72.000000   1272  58.000000    N    0.0   
145286    2016     132   1246  82.000000   1401  77.000000    N    1.0   
145287    2016     132   1277  66.000000   1345  62.000000    N    0.0   
145288    2016     132   1386  87.000000   1433  74.000000    N    0.0   

        Difference      A      H      N  
0        17.000000  False  False   True  
2        12.113395  False   True  False  
3        16.000000  False   True  False  
4        12.000000  False   True  False  
5         1.000000  False   True  False  
...            ...    ...    ...    ...  
145284   20.000000  False  False   True  
145285   14.000000  False  False   True  
145286    5.000000  False  False   True  
145287    4.000000  False  False   True  
145288   13.000000  False  False   True  

[130761 rows x 12 columns]

In [184]:
#let's drop the intial column
df.drop(columns=['Wloc'], inplace=True)
df.head()

Season  Daynum  Wteam     Wscore  Lteam     Lscore  Numot  Difference  \
0    1985      20   1228  81.000000   1328  64.000000    0.0   17.000000   
2    1985      25   1112  76.599858   1223  64.486463    0.0   12.113395   
3    1985      25   1165  70.000000   1432  54.000000    0.0   16.000000   
4    1985      25   1192  86.000000   1447  74.000000    0.0   12.000000   
5    1985      25   1218  79.000000   1337  78.000000    0.0    1.000000   

       A      H      N  
0  False  False   True  
2  False   True  False  
3  False   True  False  
4  False   True  False  
5  False   True  False

# More on extracting Rows and Columns




The bracket indexing operator is one way to extract certain columns from a dataframe.

In [185]:
df[['Wscore', 'Lscore']].head()

Wscore     Lscore
0  81.000000  64.000000
2  76.599858  64.486463
3  70.000000  54.000000
4  86.000000  74.000000
5  79.000000  78.000000

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks.

In [186]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore     Lscore
0  81.000000  64.000000
2  76.599858  64.486463
3  70.000000  54.000000
4  86.000000  74.000000
5  79.000000  78.000000

Note the difference is the return types when you use brackets and when you use double brackets.

In [113]:
type(df['Wscore'])

pandas.core.series.Series

In [114]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

In [115]:
df.head()

Season  Daynum  Wteam     Wscore  Lteam     Lscore Wloc  Numot  Difference
0    1985      20   1228  81.000000   1328  64.000000    N    0.0   17.000000
2    1985      25   1112  76.599858   1223  64.486463    H    0.0   12.113395
3    1985      25   1165  70.000000   1432  54.000000    H    0.0   16.000000
4    1985      25   1192  86.000000   1447  74.000000    H    0.0   12.000000
5    1985      25   1218  79.000000   1337  78.000000    H    0.0    1.000000

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations.

In [116]:
df[0:3]  #if you give pandas slice, it will assume your'e accessing the rows. However df[0] doesnot work (unless you have a column named 0)

Season  Daynum  Wteam     Wscore  Lteam     Lscore Wloc  Numot  Difference
0    1985      20   1228  81.000000   1328  64.000000    N    0.0   17.000000
2    1985      25   1112  76.599858   1223  64.486463    H    0.0   12.113395
3    1985      25   1165  70.000000   1432  54.000000    H    0.0   16.000000

Here's an equivalent using iloc

In [117]:
df.iloc[0:3,:]

Season  Daynum  Wteam     Wscore  Lteam     Lscore Wloc  Numot  Difference
0    1985      20   1228  81.000000   1328  64.000000    N    0.0   17.000000
2    1985      25   1112  76.599858   1223  64.486463    H    0.0   12.113395
3    1985      25   1165  70.000000   1432  54.000000    H    0.0   16.000000

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library.
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y